### PR TITLE
rgw_file: introduce new fsid and rgw_mount

### PR DESCRIPTION
**Pull request by guihecheng**
- new fsid hashed from the s3 uid, so we can get rid of the fs_inst_counster as fsid and have no more automount problems due to cross-fs access.
- new rgw_mount() with "bucket name", so we can explicitly mount a bucket (originally, we do a bucket mount by rgw_mount first then do lookup_path in nfs-ganesha, just as if we export a sub-dir mount. With this change, we have to refactor lookup_path in nfs-ganesha FSAL_RGW). This makes root_fh points to the "top" of an exported fs, a bucket or "/".

### CODE REVIEW

**Comment by Wilhelmshaven** [comment]
@guihecheng Might have conflict with your `static constexpr uint32_t FLAG_MOUNT = 0x1000;`

**Comment by mattbenjamin** [comment]
@guihecheng is there already a matching nfs-ganesha PR?  I think people may ask me to hold for 2.6, but that -should- be soon...

**Comment by guihecheng** [comment]
@mattbenjamin nfs-ganesha PR is not posted, but I've got a patch for FSAL_RGW, will raise a PR soon ;)

**Comment by mattbenjamin** [comment]
@guihecheng ok, great;  just wanted to make clear why this PR is approved but not merged (pending arrival of the matching nfs-ganesha PR)

**Comment by guihecheng** [comment]
@mattbenjamin Here is the nfs-ganesha side patch: https://review.gerrithub.io/#/c/363503/
please help view, thanks.

**Comment by mattbenjamin** [comment]
@guihecheng cool;  as well, I think this probably needs to increase the minor version in rados/rgw_file.h, as it changes semantics of rgw_mount()?

**Comment by guihecheng** [comment]
@mattbenjamin ok, sure.

File: src/rgw/rgw_file.h [hunk 1]
```diff
@@ -248,33 +248,38 @@ namespace rgw {
...
 	fh.fh_type = RGW_FS_TYPE_DIRECTORY;
 	variant_type = directory();
 	/* stat */
-	state.dev = fs_inst;
 	state.unix_mode = RGW_RWXMODE|S_IFDIR;
 	/* pointer to self */
 	fh.fh_private = this;
       }
 
+    uint64_t init_fsid(std::string& uid) {
```
Comment (mattbenjamin) [inline src/rgw/rgw_file.h]: I'm confused a bit here;  at line 276, we call it with "fsid" which is the name of the argument;  does this say that fsid == user name?  I do think that was the original intent of fs_inst() so logically it seems correct

File: src/rgw/rgw_file.h [hunk 2]
```diff
@@ -298,8 +303,8 @@ namespace rgw {
       /* save constant fhk */
       fh.fh_hk = fhk.fh_hk; /* XXX redundant in fh_hk */
 
-      /* stat */
-      state.dev = fs_inst;
+      /* inherits parent's fsid */
+      state.dev = parent->state.dev;
```
Comment (mattbenjamin) [inline src/rgw/rgw_file.h]: ok

File: src/rgw/rgw_file.h [hunk 3]
```diff
@@ -818,25 +821,23 @@ namespace rgw {
...
 	fh_lru(cct->_conf->rgw_nfs_lru_lanes,
 	       cct->_conf->rgw_nfs_lru_lane_hiwat),
 	uid(_uid), key(_user_id, _key) {
 
-      /* no bucket may be named rgw_fs_inst-(.*) */
-      fsid = RGWFileHandle::root_name + "rgw_fs_inst-" +
-	std::to_string(get_inst());
-
-      root_fh.init_rootfs(fsid /* bucket */, RGWFileHandle::root_name);
+      root_fh.init_rootfs(uid, RGWFileHandle::root_name);
```
Comment (mattbenjamin) [inline src/rgw/rgw_file.h]: cool

File: src/rgw/rgw_file.h [hunk 4]
```diff
@@ -1146,7 +1147,7 @@ namespace rgw {
 
     struct rgw_fs* get_fs() { return &fs; }
 
-    uint32_t get_inst() { return root_fh.state.dev; }
+    uint64_t get_fsid() { return root_fh.state.dev; }
```
Comment (mattbenjamin) [inline src/rgw/rgw_file.h]: ok

**Comment by guihecheng** [comment]
@mattbenjamin oh, I think rgw_mount2() is quite reasonable, I think I'll adopt this idea for the nfs-ganesha 2.6, thx.

File: src/rgw/rgw_file.h [hunk 5]
```diff
@@ -1026,7 +1041,7 @@ namespace rgw {
 	  fh_cache.insert_latched(fh, lat, RGWFileHandle::FHCache::FLAG_UNLOCK);
 	  get<1>(fhr) |= RGWFileHandle::FLAG_CREATE;
 	  /* ref parent (non-initial ref cannot fail on valid object) */
-	  if (! parent->is_root()) {
+	  if (! parent->is_mount()) {
```
Comment (guihecheng) [inline src/rgw/rgw_file.h]: @mattbenjamin iiuc, is_root() aims to check for the real "/", and is_mount() check for a root_fh which may points to the "/" or a bucket, and root_fh is not supposed to be in LRU, so ref() & unref() should adopt is_mount().

**Comment by mattbenjamin** [comment]
@guihecheng cool;  we can merge a change introducing rgw_mount2 now, and, as you say, we'd be able to use it conditionally in nfs-ganesha 2.6 (i.e., depending on the interface version of the underlying librgw_file).
** missed this (or lost it):  we also need to increment  LIBRGW_FILE_VER_EXTRA in rados/rgw_file.h

File: src/rgw/rgw_file.h (hunk 5 above)
Comment (mattbenjamin) [inline src/rgw/rgw_file.h]: ok

**Comment by guihecheng** [comment]
@mattbenjamin the rgw_mount2 is added there with LIBRGW_FILE_VER_EXTRA incremented, also, I tried to do update for old fsid just as we did for fhk, thanks.

**Comment by guihecheng** [comment]
@mattbenjamin hi, the nfs-ganesha side patch is merged recently, shall we make progress with this one ?

**Comment by mattbenjamin** [comment]
@guihecheng I think so, definitely this week or next;  I'll do some testing with it today

**Comment by mattbenjamin** [comment]
@guihecheng I've succesfully tested before and after upgrade scenario w/nfs-ganesha v2.5-stable (2.5.11), which seems to establish this is safe for the public :)

**Comment by mattbenjamin** [comment]
jenkins retest this please

**Pull request merged**